File: app/controllers/CourseAdminController.java
package controllers;

import javax.inject.Inject;

import controllers.core.BaseController;
import model.AdminSecured;
import model.StringConsts;
import model.exercisereading.ExerciseReader;
import model.user.Course;
import model.user.CourseRole;
import model.user.CourseRoleKey;
import model.user.Role;
import model.user.User;
import play.data.FormFactory;
import play.mvc.Result;
import play.mvc.Security.Authenticated;

@Authenticated(AdminSecured.class)
public class CourseAdminController extends BaseController {
  
  @Inject
  public CourseAdminController(FormFactory theFactory) {
    super(theFactory);
  }
  
  public Result addAdmin(int courseId) {
    // FIXME: AJAX!
    String userName = factory.form().bindFromRequest().get(StringConsts.NAME_NAME);
    
    User user = User.finder.byId(userName);
    if(user == null)
      return badRequest("Der Nutzer mit Name >>" + userName + "<< existiert nicht!");
    
    CourseRoleKey key = new CourseRoleKey(user.name, courseId);
    CourseRole courseRole = CourseRole.finder.byId(key);
    if(courseRole == null)
      courseRole = new CourseRole(key);
    
    courseRole.role = Role.ADMIN;
    courseRole.save();
    
    return redirect(routes.CourseAdminController.course(courseId));
  }
  
  public Result course(int id) {
    return ok(views.html.admin.course.render(getUser(), Course.finder.byId(id)));
  }
  
  public Result newCourse() {
    // FIXME: AJAX!
    String courseName = factory.form().bindFromRequest().get(StringConsts.NAME_NAME);
    
    Course course = Course.finder.all().stream().filter(c -> c.name.equals(courseName)).findAny().orElse(null);
    if(course != null)
      return badRequest("Kurs mit Namen >>" + courseName + "<< existiert bereits!");
    
    course = new Course(ExerciseReader.findMinimalNotUsedId(Course.finder));
    course.name = courseName;
    
    course.save();
    
    // Create course admin with current user
    CourseRole firstAdmin = new CourseRole(new CourseRoleKey(getUser().name, course.id));
    firstAdmin.role = Role.ADMIN;
    firstAdmin.save();
    
    return redirect(routes.AdminController.index());
  }
  
  public Result newCourseForm() {
    return ok(views.html.admin.newCourseForm.render(getUser()));
  }
}

File: app/controllers/CourseAdminController.scala
package controllers

import javax.inject._

import model.core.Repository
import play.api.db.slick.{DatabaseConfigProvider, HasDatabaseConfigProvider}
import play.api.mvc.{AbstractController, ControllerComponents, EssentialAction}
import slick.jdbc.JdbcProfile

import scala.concurrent.ExecutionContext

class CourseAdminController @Inject()(cc: ControllerComponents, val dbConfigProvider: DatabaseConfigProvider, val repository: Repository)(implicit ec: ExecutionContext)
  extends AbstractController(cc) with HasDatabaseConfigProvider[JdbcProfile] with Secured {

  override protected val adminRightsRequired: Boolean = true

  // FIXME: AJAX!
  def addAdmin(courseId: Int): EssentialAction = withUser { _ =>
    implicit request =>
      //      val userName = singleStrForm(NAME_NAME).bindFromRequest.get.str
      //      Option(User.finder.byId(userName)) match {
      //        case None => BadRequest(s"Der Nutzer mit dem Namen $userName existiert nicht!")
      //        case Some(userToChange) =>
      //          val key = new CourseRoleKey(userToChange.name, courseId)
      //          val courseRole = Option(CourseRole.finder.byId(key)).getOrElse(new CourseRole(key))
      //
      //          courseRole.role = Role.ADMIN
      //          courseRole.save()
      //
      //          Redirect(routes.CourseAdminController.course(courseId))
      //      }
      Ok("TODO!")
  }

  def course(id: Int): EssentialAction = withUser { _ =>
    implicit request =>
      //      Ok(views.html.admin.course.render(user, Course.finder.byId(id)))

      Ok("TODO!")
  }

  // FIXME: AJAX!
  def newCourse(): EssentialAction = withUser { _ =>
    implicit request =>
      //      val courseName = singleStrForm(NAME_NAME).bindFromRequest.get.str
      //      Course.finder.all.asScala.find(_.name == courseName) match {
      //        case Some(_) => BadRequest(s"Kurs mit Namen $courseName existiert bereits!")
      //        case None =>
      //          val course = new Course(courseName)
      //
      //          course.save()
      //
      //          // Create course admin with current user
      //          val firstAdmin = new CourseRole(new CourseRoleKey(user.name, course.id))
      //          firstAdmin.role = Role.ADMIN
      //          firstAdmin.save()
      //
      //          Redirect(routes.AdminController.activityDrawingIndex())
      //      }
      Ok("TODO!")
  }

  def newCourseForm(): EssentialAction = withUser { user => implicit request => Ok(views.html.admin.newCourseForm.render(user)) }
}
